File: packages/clients/ios/Sources/DoubloonChecker/EvmChecker.swift
import Foundation
import CryptoKit

/// Lightweight EVM entitlement checker using direct JSON-RPC.
///
/// Calls the Doubloon Solidity contract's `view` functions via `eth_call`.
/// Zero gas cost, no wallet required, no server needed.
///
/// ```swift
/// let checker = EvmChecker(
///     rpcUrl: URL(string: "https://eth.llamarpc.com")!,
///     contractAddress: "0xYourDoubloonContract..."
/// )
///
/// let result = try await checker.checkEntitlement(
///     productId: "a7f3c9...",
///     wallet: "0xUserAddress..."
/// )
///
/// if result.entitled {
///     // Grant access
/// }
/// ```
public actor EvmChecker {
    private let rpcUrl: URL
    private let contractAddress: String
    private let rpc: RpcClient

    public init(rpcUrl: URL, contractAddress: String) {
        self.rpcUrl = rpcUrl
        self.contractAddress = contractAddress
        self.rpc = RpcClient()
    }

    /// Check if a wallet holds an active entitlement.
    public func checkEntitlement(productId: String, wallet: String) async throws -> EntitlementCheck {
        let entitlement = try await getEntitlement(productId: productId, wallet: wallet)
        return DoubloonChecker.checkEntitlement(entitlement)
    }

    /// Quick boolean check using the contract's `isEntitled` view function.
    public func isEntitled(productId: String, wallet: String) async throws -> Bool {
        let calldata = EvmAbi.encodeIsEntitled(productId: productId, user: wallet)
        let result: String = try await ethCall(data: calldata)
        return EvmAbi.decodeBool(result)
    }

    /// Fetch the full entitlement record, or nil if not found.
    public func getEntitlement(productId: String, wallet: String) async throws -> Entitlement? {
        let calldata = EvmAbi.encodeGetEntitlement(productId: productId, user: wallet)
        let result: String = try await ethCall(data: calldata)

        let clean = result.hasPrefix("0x") ? String(result.dropFirst(2)) : result
        guard clean.count >= 64 else { return nil }

        let raw = EvmAbi.decodeGetEntitlement(result)
        guard raw.exists else { return nil }

        return raw.toEntitlement()
    }

    /// Fetch product metadata, or nil if not found.
    public func getProduct(productId: String) async throws -> Product? {
        let calldata = EvmAbi.encodeGetProduct(productId: productId)
        let result: String = try await ethCall(data: calldata)

        let clean = result.hasPrefix("0x") ? String(result.dropFirst(2)) : result
        guard clean.count >= 64 else { return nil }

        let raw = EvmAbi.decodeGetProduct(result)
        guard raw.exists else { return nil }

        return raw.toProduct()
    }

    /// Batch check multiple products.
    public func checkEntitlements(
        productIds: [String],
        wallet: String
    ) async throws -> [String: EntitlementCheck] {
        var checks: [String: EntitlementCheck] = [:]
        // Run checks concurrently
        try await withThrowingTaskGroup(of: (String, EntitlementCheck).self) { group in
            for pid in productIds {
                group.addTask {
                    let check = try await self.checkEntitlement(productId: pid, wallet: wallet)
                    return (pid, check)
                }
            }
            for try await (pid, check) in group {
                checks[pid] = check
            }
        }
        return checks
    }

    // MARK: - Private

    private func ethCall(data: String) async throws -> String {
        let params: [[String: String]] = [
            ["to": contractAddress, "data": "0x" + data],
        ]
        return try await rpc.call(
            url: rpcUrl,
            method: "eth_call",
            params: params + ["latest"] as [Any]
        )
    }
}

// MARK: - ABI Encoding/Decoding

enum EvmAbi {
    // Function selectors (first 4 bytes of keccak256 of signature)
    static let selectorIsEntitled = "2b1c1e9f"
    static let selectorGetEntitlement = "fdb60e41"
    static let selectorGetProduct = "a3e76c0f"

    static func encodeIsEntitled(productId: String, user: String) -> String {
        selectorIsEntitled + encodeBytes32(productId) + encodeAddress(user)
    }

    static func encodeGetEntitlement(productId: String, user: String) -> String {
        selectorGetEntitlement + encodeBytes32(productId) + encodeAddress(user)
    }

    static func encodeGetProduct(productId: String) -> String {
        selectorGetProduct + encodeBytes32(productId)
    }

    static func decodeBool(_ data: String) -> Bool {
        let clean = data.hasPrefix("0x") ? String(data.dropFirst(2)) : data
        guard clean.count >= 64 else { return false }
        return clean.suffix(1) != "0"
    }

    // MARK: - Encoding helpers

    private static func encodeBytes32(_ hex: String) -> String {
        let clean = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        return String(repeating: "0", count: max(0, 64 - clean.count)) + clean
    }

    private static func encodeAddress(_ addr: String) -> String {
        let clean = (addr.hasPrefix("0x") ? String(addr.dropFirst(2)) : addr).lowercased()
        return String(repeating: "0", count: max(0, 64 - clean.count)) + clean
    }

    // MARK: - Decoding

    struct RawEntitlement {
        let productId: String
        let user: String
        let grantedAt: UInt64
        let expiresAt: Int64
        let autoRenew: Bool
        let source: UInt8
        let sourceId: String
        let active: Bool
        let revokedAt: UInt64
        let revokedBy: String
        let exists: Bool

        func toEntitlement() -> Entitlement {
            let zeroAddr = "0x0000000000000000000000000000000000000000"
            return Entitlement(
                productId: productId,
                user: user,
                grantedAt: Date(timeIntervalSince1970: TimeInterval(grantedAt)),
                expiresAt: expiresAt == 0 ? nil : Date(timeIntervalSince1970: TimeInterval(expiresAt)),
                autoRenew: autoRenew,
                source: EntitlementSource(rawValue: source) ?? .platform,
                sourceId: sourceId,
                active: active,
                revokedAt: revokedAt == 0 ? nil : Date(timeIntervalSince1970: TimeInterval(revokedAt)),
                revokedBy: revokedBy == zeroAddr ? nil : revokedBy
            )
        }
    }

    struct RawProduct {
        let creator: String
        let productId: String
        let name: String
        let metadataUri: String
        let createdAt: UInt64
        let updatedAt: UInt64
        let active: Bool
        let frozen: Bool
        let entitlementCount: UInt64
        let delegateCount: UInt16
        let defaultDuration: Int64
        let exists: Bool

        func toProduct() -> Product {
            Product(
                creator: creator,
                productId: productId,
                name: name,
                metadataUri: metadataUri,
                createdAt: Date(timeIntervalSince1970: TimeInterval(createdAt)),
                updatedAt: Date(timeIntervalSince1970: TimeInterval(updatedAt)),
                active: active,
                frozen: frozen,
                entitlementCount: entitlementCount,
                delegateCount: delegateCount,
                defaultDuration: defaultDuration
            )
        }
    }

    static func decodeGetEntitlement(_ data: String) -> RawEntitlement {
        let r = SlotReader(data)
        // First slot is offset to tuple data
        let tupleOffset = r.readUint()
        r.seek(to: Int(tupleOffset))

        let productId = r.readHex(64)
        let user = r.readAddress()
        let grantedAt = r.readUint64()
        let expiresAt = r.readInt64()
        let autoRenew = r.readBool()
        let source = UInt8(r.readUint())
        let sourceIdOffset = r.readUint()
        let active = r.readBool()
        let revokedAt = r.readUint64()
        let revokedBy = r.readAddress()
        let exists = r.readBool()

        let sourceId = r.readDynamicString(at: Int(tupleOffset) + Int(sourceIdOffset))

        return RawEntitlement(
            productId: productId, user: user,
            grantedAt: grantedAt, expiresAt: expiresAt,
            autoRenew: autoRenew, source: source, sourceId: sourceId,
            active: active, revokedAt: revokedAt, revokedBy: revokedBy, exists: exists
        )
    }

    static func decodeGetProduct(_ data: String) -> RawProduct {
        let r = SlotReader(data)
        let tupleOffset = r.readUint()
        r.seek(to: Int(tupleOffset))

        let creator = r.readAddress()
        let productId = r.readHex(64)
        let nameOffset = r.readUint()
        let metadataUriOffset = r.readUint()
        let createdAt = r.readUint64()
        let updatedAt = r.readUint64()
        let active = r.readBool()
        let frozen = r.readBool()
        let entitlementCount = r.readUint64()
        let delegateCount = UInt16(r.readUint())
        let defaultDuration = r.readInt64()
        let exists = r.readBool()

        let name = r.readDynamicString(at: Int(tupleOffset) + Int(nameOffset))
        let metadataUri = r.readDynamicString(at: Int(tupleOffset) + Int(metadataUriOffset))

        return RawProduct(
            creator: creator, productId: productId,
            name: name, metadataUri: metadataUri,
            createdAt: createdAt, updatedAt: updatedAt,
            active: active, frozen: frozen,
            entitlementCount: entitlementCount, delegateCount: delegateCount,
            defaultDuration: defaultDuration, exists: exists
        )
    }
}

// MARK: - Slot Reader

private class SlotReader {
    private let hex: String
    private var pos: Int = 0

    init(_ data: String) {
        self.hex = data.hasPrefix("0x") ? String(data.dropFirst(2)) : data
    }

    func seek(to byteOffset: Int) {
        pos = byteOffset * 2
    }

    func readSlot() -> String {
        let start = hex.index(hex.startIndex, offsetBy: pos)
        let end = hex.index(start, offsetBy: 64)
        let slot = String(hex[start..<end])
        pos += 64
        return slot
    }

    func readUint() -> UInt64 {
        let slot = readSlot()
        return UInt64(slot, radix: 16) ?? 0
    }

    func readUint64() -> UInt64 { readUint() }

    func readInt64() -> Int64 {
        let val = readUint()
        if val > UInt64(Int64.max) {
            return Int64(bitPattern: val)
        }
        return Int64(val)
    }

    func readBool() -> Bool { readUint() != 0 }

    func readAddress() -> String {
        let slot = readSlot()
        return "0x" + String(slot.suffix(40))
    }

    func readHex(_ chars: Int) -> String {
        let slot = readSlot()
        return String(slot.prefix(chars))
    }

    func readDynamicString(at byteOffset: Int) -> String {
        let startHex = byteOffset * 2
        let lenStart = hex.index(hex.startIndex, offsetBy: startHex)
        let lenEnd = hex.index(lenStart, offsetBy: 64)
        let len = Int(String(hex[lenStart..<lenEnd]), radix: 16) ?? 0
        guard len > 0 else { return "" }

        let strStart = hex.index(lenEnd, offsetBy: 0)
        let strEnd = hex.index(strStart, offsetBy: len * 2)
        let strHex = String(hex[strStart..<strEnd])

        var bytes: [UInt8] = []
        var i = strHex.startIndex
        while i < strHex.endIndex {
            let next = strHex.index(i, offsetBy: 2)
            bytes.append(UInt8(strHex[i..<next], radix: 16) ?? 0)
            i = next
        }

        return String(bytes: bytes, encoding: .utf8) ?? ""
    }
}

File: packages/clients/android/src/main/kotlin/com/doubloon/checker/EvmChecker.kt
package com.doubloon.checker

import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import org.json.JSONArray
import org.json.JSONObject
import java.util.Date

/**
 * Lightweight EVM entitlement checker using direct JSON-RPC.
 *
 * Calls the Doubloon Solidity contract's `view` functions via `eth_call`.
 * Zero gas cost, no wallet required, no server needed.
 *
 * ```kotlin
 * val checker = EvmChecker(
 *     rpcUrl = "https://eth.llamarpc.com",
 *     contractAddress = "0xYourDoubloonContract..."
 * )
 *
 * val result = checker.checkEntitlement("a7f3c9...", "0xUserAddress...")
 * if (result.entitled) {
 *     // Grant access
 * }
 * ```
 */
class EvmChecker(
    private val rpcUrl: String,
    private val contractAddress: String,
    private val rpc: RpcClient = RpcClient(),
) {
    /**
     * Check if a wallet holds an active entitlement.
     */
    suspend fun checkEntitlement(productId: String, wallet: String): EntitlementCheck {
        val entitlement = getEntitlement(productId, wallet)
        return com.doubloon.checker.checkEntitlement(entitlement)
    }

    /**
     * Quick boolean check using the contract's `isEntitled` view function.
     */
    suspend fun isEntitled(productId: String, wallet: String): Boolean {
        val calldata = EvmAbi.encodeIsEntitled(productId, wallet)
        val result = ethCall(calldata)
        return EvmAbi.decodeBool(result)
    }

    /**
     * Fetch the full entitlement record, or null if not found.
     */
    suspend fun getEntitlement(productId: String, wallet: String): Entitlement? {
        val calldata = EvmAbi.encodeGetEntitlement(productId, wallet)
        val result = ethCall(calldata)

        val clean = if (result.startsWith("0x")) result.substring(2) else result
        if (clean.length < 64) return null

        val raw = EvmAbi.decodeGetEntitlement(result)
        if (!raw.exists) return null

        return raw.toEntitlement()
    }

    /**
     * Fetch product metadata, or null if not found.
     */
    suspend fun getProduct(productId: String): Product? {
        val calldata = EvmAbi.encodeGetProduct(productId)
        val result = ethCall(calldata)

        val clean = if (result.startsWith("0x")) result.substring(2) else result
        if (clean.length < 64) return null

        val raw = EvmAbi.decodeGetProduct(result)
        if (!raw.exists) return null

        return raw.toProduct()
    }

    /**
     * Batch check multiple products.
     */
    suspend fun checkEntitlements(
        productIds: List<String>,
        wallet: String,
    ): Map<String, EntitlementCheck> = coroutineScope {
        productIds.map { pid ->
            async { pid to checkEntitlement(pid, wallet) }
        }.awaitAll().toMap()
    }

    private suspend fun ethCall(data: String): String {
        val params = JSONArray().apply {
            put(JSONObject().apply {
                put("to", contractAddress)
                put("data", "0x$data")
            })
            put("latest")
        }
        return rpc.callForString(rpcUrl, "eth_call", params)
    }
}

// MARK: - ABI Encoding/Decoding

object EvmAbi {
    // Function selectors
    private const val SELECTOR_IS_ENTITLED = "2b1c1e9f"
    private const val SELECTOR_GET_ENTITLEMENT = "fdb60e41"
    private const val SELECTOR_GET_PRODUCT = "a3e76c0f"

    fun encodeIsEntitled(productId: String, user: String): String =
        SELECTOR_IS_ENTITLED + encodeBytes32(productId) + encodeAddress(user)

    fun encodeGetEntitlement(productId: String, user: String): String =
        SELECTOR_GET_ENTITLEMENT + encodeBytes32(productId) + encodeAddress(user)

    fun encodeGetProduct(productId: String): String =
        SELECTOR_GET_PRODUCT + encodeBytes32(productId)

    fun decodeBool(data: String): Boolean {
        val clean = if (data.startsWith("0x")) data.substring(2) else data
        if (clean.length < 64) return false
        return clean.substring(0, 64).trimStart('0').isNotEmpty()
    }

    private fun encodeBytes32(hex: String): String {
        val clean = if (hex.startsWith("0x")) hex.substring(2) else hex
        return clean.padStart(64, '0')
    }

    private fun encodeAddress(addr: String): String {
        val clean = (if (addr.startsWith("0x")) addr.substring(2) else addr).lowercase()
        return clean.padStart(64, '0')
    }

    data class RawEntitlement(
        val productId: String,
        val user: String,
        val grantedAt: Long,
        val expiresAt: Long,
        val autoRenew: Boolean,
        val source: Int,
        val sourceId: String,
        val active: Boolean,
        val revokedAt: Long,
        val revokedBy: String,
        val exists: Boolean,
    ) {
        fun toEntitlement(): Entitlement {
            val zeroAddr = "0x0000000000000000000000000000000000000000"
            return Entitlement(
                productId = productId,
                user = user,
                grantedAt = Date(grantedAt * 1000),
                expiresAt = if (expiresAt == 0L) null else Date(expiresAt * 1000),
                autoRenew = autoRenew,
                source = EntitlementSource.fromValue(source),
                sourceId = sourceId,
                active = active,
                revokedAt = if (revokedAt == 0L) null else Date(revokedAt * 1000),
                revokedBy = if (revokedBy == zeroAddr) null else revokedBy,
            )
        }
    }

    data class RawProduct(
        val creator: String,
        val productId: String,
        val name: String,
        val metadataUri: String,
        val createdAt: Long,
        val updatedAt: Long,
        val active: Boolean,
        val frozen: Boolean,
        val entitlementCount: Long,
        val delegateCount: Int,
        val defaultDuration: Long,
        val exists: Boolean,
    ) {
        fun toProduct(): Product = Product(
            creator = creator,
            productId = productId,
            name = name,
            metadataUri = metadataUri,
            createdAt = Date(createdAt * 1000),
            updatedAt = Date(updatedAt * 1000),
            active = active,
            frozen = frozen,
            entitlementCount = entitlementCount,
            delegateCount = delegateCount,
            defaultDuration = defaultDuration,
        )
    }

    fun decodeGetEntitlement(data: String): RawEntitlement {
        val r = SlotReader(data)
        val tupleOffset = r.readUint().toInt()
        r.seek(tupleOffset)

        val productId = r.readHex(64)
        val user = r.readAddress()
        val grantedAt = r.readUint()
        val expiresAt = r.readInt64()
        val autoRenew = r.readBool()
        val source = r.readUint().toInt()
        val sourceIdOffset = r.readUint().toInt()
        val active = r.readBool()
        val revokedAt = r.readUint()
        val revokedBy = r.readAddress()
        val exists = r.readBool()

        val sourceId = r.readDynamicString(tupleOffset + sourceIdOffset)

        return RawEntitlement(productId, user, grantedAt, expiresAt, autoRenew, source, sourceId, active, revokedAt, revokedBy, exists)
    }

    fun decodeGetProduct(data: String): RawProduct {
        val r = SlotReader(data)
        val tupleOffset = r.readUint().toInt()
        r.seek(tupleOffset)

        val creator = r.readAddress()
        val productId = r.readHex(64)
        val nameOffset = r.readUint().toInt()
        val metadataUriOffset = r.readUint().toInt()
        val createdAt = r.readUint()
        val updatedAt = r.readUint()
        val active = r.readBool()
        val frozen = r.readBool()
        val entitlementCount = r.readUint()
        val delegateCount = r.readUint().toInt()
        val defaultDuration = r.readInt64()
        val exists = r.readBool()

        val name = r.readDynamicString(tupleOffset + nameOffset)
        val metadataUri = r.readDynamicString(tupleOffset + metadataUriOffset)

        return RawProduct(creator, productId, name, metadataUri, createdAt, updatedAt, active, frozen, entitlementCount, delegateCount, defaultDuration, exists)
    }

    private class SlotReader(data: String) {
        private val hex = if (data.startsWith("0x")) data.substring(2) else data
        private var pos = 0

        fun seek(byteOffset: Int) {
            pos = byteOffset * 2
        }

        fun readSlot(): String {
            val slot = hex.substring(pos, pos + 64)
            pos += 64
            return slot
        }

        fun readUint(): Long = java.lang.Long.parseUnsignedLong(readSlot().takeLast(16), 16)

        fun readInt64(): Long {
            val slot = readSlot()
            val unsigned = java.lang.Long.parseUnsignedLong(slot.takeLast(16), 16)
            return unsigned // Java long is already signed 64-bit
        }

        fun readBool(): Boolean = readUint() != 0L

        fun readAddress(): String = "0x" + readSlot().takeLast(40)

        fun readHex(chars: Int): String = readSlot().take(chars)

        fun readDynamicString(byteOffset: Int): String {
            val startHex = byteOffset * 2
            val len = hex.substring(startHex, startHex + 64).trimStart('0').ifEmpty { "0" }.toLong(16).toInt()
            if (len == 0) return ""
            val strHex = hex.substring(startHex + 64, startHex + 64 + len * 2)
            return ByteArray(len) { i ->
                strHex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
            }.toString(Charsets.UTF_8)
        }
    }
}
